Decouple identity policy from endpoint policy

joestringer opened the pull request:
Decouple the resolving of policy for a particular identity from the endpoint-specific distillation of that policy. The `SelectorPolicy` deals with all rule selectors that apply to a particular identity (of an endpoint), and this performs merging of the policy. The `EndpointPolicy` applies that partially-resolved `SelectorPolicy` to a specific set of identity peers in context of generating the endpoint BPF policy map contents.

As part of this, rework the rule selector cache to be based upon identities rather than endpoints. This makes use of the recently-added identitymanager (#7616) to determine when an identity can be safely flushed from the cache.

Related: #7515 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7542)
<!-- Reviewable:end -->

joestringer committed:
policy: Split SelectorPolicy from EndpointPolicy

* SelectorPolicy represents the resolved policy for an Identity, still
  in terms of the label selectors and so on, but with rules merged into a
  common representation. These are oblivious to concrete known identities.

* EndpointPolicy represents the SelectorPolicy which has been
  "distilled" down to the sets of datapath fields that are required to
  implement the policy. This is generated from a set of identities that
  are currently known by Cilium.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
policy/rule: Convert selection cache to identity

Convert the rule selection cache from keying by endpoint ID to keying by
security identity. This simplifies the cache and allows it to be shared
by multiple endpoints using the same SelectorPolicy, and also removes
the last remaining piece of endpoint-specific data from the
SelectorPolicy.

We hook this into the recently-added identitymanager such that when
identities are no longer used locally, the selector cache will trigger
an event that clears references to that identity from the policy
repository. This can reduce contention on the policy lock in cases where
endpoints with the same identity are frequently added and removed.

Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on pkg/policy/resolve.go:
```diff
@@ -18,14 +18,16 @@ import (
+// IdentityPolicy is a structure which contains the resolved policy across all
+// layers (L3, L4, and L7), but still determined in terms of EndpointSelectors.
+type IdentityPolicy struct {
 	// ID is the node-local identifier of this EndpointPolicy.
```
s/EndpointPolicy/IdentityPolicy

joestringer replied:
This is removed in a subsequent commit.

ianvernon commented on pkg/policy/resolve.go:
```diff
@@ -86,6 +102,85 @@ func getSecurityIdentities(labelsMap cache.IdentityCache, selector *api.Endpoint
+	if p.IngressPolicyEnabled {
+		for identity, labels := range identityCache {
+			ingressCtx.From = labels
+			egressCtx.To = labels
```
I think this line, while harmless, can be removed - the ingressContext is the only thing which is being evaluated here. This is in the old code as well.

joestringer replied:
Yeah, I kept it the same just so that the individual commits are direct refactors. I can look into this.

ianvernon commented on pkg/policy/resolve.go:
```diff
@@ -86,6 +102,85 @@ func getSecurityIdentities(labelsMap cache.IdentityCache, selector *api.Endpoint
+		DeniedEgressIdentities:  cache.IdentityCache{},
+	}
+
+	labels := policyOwner.GetSecurityIdentity().LabelArray
```
There is a possibility that the returned value from `GetSecurityIdentity()` could be nil since it returns a pointer.

joestringer replied:
Good spotting. This actually highlights that this function shouldn't receive a `PolicyOwner` (which represents an endpoint), but instead it should receive the identity directly. Changing this over should tidy up the interface quite nicely.

joestringer replied:
On closer inspection, we're still depending on endpoint-specific state here for the moment so can't switch it over. Will be able to tidy this up when #6921 lands.

ianvernon commented on daemon/daemon.go:
```diff
@@ -1604,8 +1604,21 @@ func (d *Daemon) GetNodeSuffix() string {
-	return d.policy.RemoveEndpointIDFromRuleCaches(id)
+// ReleaseIdentity asynchronously removes references to the specified identity
+// from the rules in the daemon's policy repository.
+func (d *Daemon) ReleaseIdentity(identity *identity.Identity) {
```
Nit: this function naming collides with that of the allocator - I think this should be changed to something like `RemoveIdentityFromPolicyCaches` - this is verbose, but more clear.

joestringer replied:
This should be resolved when combining with #7539 .

joestringer replied:
FYI these bits were stripped out when I rebased against identitymanager (#7616).

ianvernon commented on pkg/policy/repository_test.go:
```diff
@@ -131,138 +131,82 @@ func (ds *PolicyTestSuite) TestComputePolicyEnforcementAndRules(c *C) {
 	}
-	dummyEndpointID := uint16(7777)
 
-	convertedFooIngressRule1 := &rule{
```
I think we should still check that the caches are updated in this test. Why can't the testing coverage be the same as before with ensuring that the `ruleMetadata` is updated correctly as well?

EDIT - this is covered by other unit tests, so this is OK. Scopes the test to ensure that the rules that are returned are correct, which is OK IMO.

joestringer replied:
Yeah, when I looked at these tests my primary complaint was that we were testing a bunch of internal details which should be covered by something that uses the object functions to determine that the state is correct. Any type of modification like what we do in this PR, including especially private member changes would mean that these test have to be modified. I didn't get an evaluation back from coveralls yet but I don't think that these particular changes should reduce the coverage.

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -37,29 +37,21 @@ type ruleMetadata struct {
-	AllEndpoints map[uint16]struct{}
+	// IdentitySelected is a cache that maps from an identity to whether
+	// this rule selects that identity.
+	IdentitySelected map[*identity.Identity]bool
```
Nice, this is much cleaner than what I had before!

ianvernon reviewed (approved):
Nothing blocking - please add the nil check for `GetSecurityIdentity` in `pkg/policy/resolve`, though as a follow-up commit. Great work!

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -217,6 +229,14 @@ func AllocateIdentity(ctx context.Context, lbls labels.Labels) (*identity.Identi
 // identity again. This function may result in kvstore operations.
 // After the last user has released the ID, the returned lastUse value is true.
 func Release(ctx context.Context, id *identity.Identity) (bool, error) {
+	lastUse, err := release(ctx, id)
```
One question - what is the scope of `lastUse` - is it global or local to this node?

joestringer replied:
Ah, I believe that this is the last local reference to this identity, but I seem to recall @jrajahalme talking about this not having the semantics that one might expect---specifically, that it will return true when removing the final reference to an identity, but then if the identity is referenced again after this and unreferenced again, it would not return true this time. This part may end up going away when I rebase against #7539 .

joestringer commented on pkg/identity/identity.go:
```diff
@@ -140,6 +140,11 @@ func (id *Identity) IsWellKnown() bool {
 
+// IsGlobal returns true if the identity is backed by a global identity.
+func (id *Identity) IsGlobal() bool {
+	return RequiresGlobalIdentity(id.Labels)
```
Discussed offline that this kind of functionality could act upon `NumericIdentity` instead to simplify merge with #7539. Alternatively #7616 may provide a better alternative without needing to interact at this layer.

joestringer replied:
Will look into rebasing against #7616.

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -18,17 +18,16 @@ import (
-
+// IdentityPolicy is a structure which contains the resolved policy across all
+// layers (L3, L4, and L7), but still determined in terms of EndpointSelectors.
+type IdentityPolicy struct {
```
The more I think about this, the more I think that `SelectorPolicy` is a more appropriate name for this level of abstraction. It *is* the policy for a particular Identity, and we will generate it when an identity is used by a local endpoint.... however, changes to remote identities do not result in changes to this structure. Rather, changes to policy rule endpoint selectors for allowing interaction with network peers change this structure. Furthermore, the contributing aspect of policy to this structure is whether the policy rules *select* endpoints with this identity.

Naming as `SelectorPolicy` would avoid confusion on this front by putting more distance between identity interactions and this structure.

joestringer committed:
daemon: Don't populate rule selector cache on restore

Let the policy rule selector cache be lazily populated upon demand
rather than blocking on this during restore for each endpoint.

This simplifies upcoming changes.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
identitymanager: Support subscribing to events

Add a simple callback mechanism to allow observers to be notified when
an identity becomes used or no longer used by endpoints that are local
to the node.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
identitymanager: Simplify labels in test

These will be reused by an upcoming commit, factor them into variables
and simplify.

Signed-off-by: Joe Stringer <joe@cilium.io>

ianvernon commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -143,6 +147,12 @@ func (idm *IdentityManager) GetIdentityModels() []*models.IdentityEndpoints {
+func (idm *IdentityManager) subscribe(o Observer) {
+	idm.mutex.Lock()
+	defer idm.mutex.Unlock()
+	idm.observers = append(idm.observers, o)
```
I'd prefer to use `map[Observer]struct{}` to ensure that we can't update a given observer multiple times, which could occur if we call `Subscribe` multiple times for the same observer. However, this ensures that an `Observer` is always a comparable type, which we can't guarantee at compile-time.

joestringer replied:
This occurred to me, but with the minimal usage there is in this PR (exactly one static subscribe) I didn't see a strong motivation just yet. If you think we're likely to miss problems of this kind when extending this code in future then maybe I better make this change now.

joestringer replied:
Addressed.

ianvernon commented on pkg/policy/rules.go:
```diff
@@ -381,16 +381,14 @@ func (rules ruleSlice) updateEndpointsCaches(ep Endpoint, epIDSet *IDSet) bool {
 		return
 	}
 
-	id := ep.GetID16()
```
does the `Endpoint` interface need `GetID16` as a required function anymore?

joestringer replied:
It's still used to collect a group of endpoint IDs that need regenerating based upon policy changes---see `UpdateRulesEndpointsCaches()`.

joestringer commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -122,6 +124,9 @@ func (idm *IdentityManager) remove(identity *identity.Identity) {
 	idMeta.refCount--
 	if idMeta.refCount == 0 {
 		delete(idm.identities, identity.ID)
+		for o, _ := range idm.observers {
```
^^^ Need to resolve this, causes `make` failure.

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -639,33 +636,32 @@ func (p *Repository) GetRulesList() *models.Policy {
-		DeniedEgressIdentities:  cache.IdentityCache{},
-	}
+// ResolvePolicyLocked returns the EndpointPolicy for the provided set of
+// labels against the set of rules in the repository, and the provided set of
```
"EndpointPolicy" -> "SelectorPolicy"

joestringer replied:
Fixed by #7617

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -639,33 +636,32 @@ func (p *Repository) GetRulesList() *models.Policy {
+		EgressPolicyEnabled:  egressEnabled,
+	}
 	calculatedPolicy.IngressPolicyEnabled = ingressEnabled
 	calculatedPolicy.EgressPolicyEnabled = egressEnabled
```
These two lines seem unnecessary now.

joestringer replied:
I'll strip these out in #7617

jrajahalme commented on pkg/policy/rule.go:
```diff
@@ -37,29 +37,21 @@ type ruleMetadata struct {
 	return &ruleMetadata{
-		EndpointsSelected: make(map[uint16]*identity.Identity),
-		AllEndpoints:      make(map[uint16]struct{}),
+		IdentitySelected: make(map[*identity.Identity]bool),
```
Does it matter that two Identity instances could describe the same security identity? If yes then maybe use the numeric identity as the map key instead?

joestringer replied:
It doesn't matter, this is purely a cache and if an entry is not found, it will be populated.

Do you know if this happens?

jrajahalme reviewed (commented):
Looking good, some nits only.
